File: javascript/micm/conditions.js
import { GAS_CONSTANT } from './utils.js';

/**
 * Environmental conditions for a single grid cell.
 *
 * If `air_density` is not provided and both `temperature` and `pressure` are
 * given, air density is computed from the Ideal Gas Law: ``P / (R * T)``.
 */
export class Conditions {
  /**
   * @param {Object} [options]
   * @param {number|null} [options.temperature=null] - Temperature in Kelvin
   * @param {number|null} [options.pressure=null] - Pressure in Pascals
   * @param {number|null} [options.air_density=null] - Air number density in mol m⁻³;
   *   computed from the Ideal Gas Law when `null` and both temperature and pressure are set
   */
  constructor({ temperature = null, pressure = null, air_density = null } = {}) {
    this.temperature = temperature;
    this.pressure = pressure;
    if (air_density !== null) {
      this.air_density = air_density;
    } else if (temperature !== null && pressure !== null) {
      this.air_density = 1.0 / ((GAS_CONSTANT * temperature) / pressure);
    } else {
      this.air_density = null;
    }
  }
}

File: julia/src/micm/conditions.jl
# Copyright (C) 2023-2026 University Corporation for Atmospheric Research
# SPDX-License-Identifier: Apache-2.0

"""
    Conditions

Environmental conditions for a grid cell.

If `air_density` is not provided and both `temperature` and `pressure` are given,
air density is calculated from the Ideal Gas Law.

# Fields
- `temperature::Float64`: Temperature in Kelvin
- `pressure::Float64`: Pressure in Pascals
- `air_density::Float64`: Air density in mol m⁻³
"""
mutable struct Conditions
    temperature::Float64
    pressure::Float64
    air_density::Float64
end

function Conditions(; temperature::Real=0.0, pressure::Real=0.0, air_density::Union{Real, Nothing}=nothing)
    if air_density === nothing
        if temperature > 0.0 && pressure > 0.0
            air_density = pressure / (GAS_CONSTANT * temperature)
        else
            air_density = 0.0
        end
    end
    return Conditions(Float64(temperature), Float64(pressure), Float64(air_density))
end
